TimeSeries: Render thresholds as lines and areas

torkelo opened the pull request:
An initial attempt at rendering thresholds for TimeSeries/GraphNG   (based on https://github.com/grafana/grafana/pull/33215 branch to avoid later conflicts) 

Adds a GraphThresholdsConfig with a mode option  (None, Line, Area), this is currently an object so we can support more related options like threshold opacity. Could technically be defined manually with the threshold color, but then tricky if you want to combine line+area.  

For the areas you have define the full area using multiple thresholds (going to infinity automatically is not supported as that would make area bands impossible). So to make an area go orange from 60 - 80 would require a 60 threshold with color orange, the 80 thresholds can be red but to have it render an area a user needs to add an additional threshold say (100) to get a red area from 80 to 100. 

![Screenshot from 2021-04-25 09-25-16](https://user-images.githubusercontent.com/10999/115984733-f3e6a580-a5a8-11eb-867a-7fb067139887.png)
![Screenshot from 2021-04-25 09-25-11](https://user-images.githubusercontent.com/10999/115984734-f47f3c00-a5a8-11eb-86b9-381da89c17b4.png)

torkelo committed:
TimeSeries: First take on uplot thresholds rendering

leeoniya commented on packages/grafana-ui/src/components/uPlot/config/UPlotConfigBuilder.ts:
```diff
@@ -36,6 +39,13 @@ export class UPlotConfigBuilder {
+  addThresholds(options: UPlotThresholdOptions) {
+    if (!this.thresholds[options.scaleKey]) {
+      this.thresholds[options.scaleKey] = options;
+      this.addHook('draw', getThresholdsDrawHook(options));
```
tbe `draw` hook fires last, so you'll be painting thresholds on top of everything else. not sure if this is desired behavior. you may consider an earlier hook like `drawAxes` (to paint over grid) or `drawClear` to paint under everything. or even add and option for users to choose a layer ;)

dprokop replied:
I'm confused. This is with current impementation (`draw`):

![image](https://user-images.githubusercontent.com/2376619/116074340-e0fbd000-a691-11eb-9363-bec880791e98.png)

This is with `drawAxes`:
![image](https://user-images.githubusercontent.com/2376619/116074217-b6aa1280-a691-11eb-9cef-6c18269dfc1d.png)

And this is with `drawClear`:
![image](https://user-images.githubusercontent.com/2376619/116074291-cde90000-a691-11eb-9739-df5896414770.png)


Can't see any difference 🤔

leeoniya replied:
you have to squint, but you can see that the yellow line is faded by the threshold band in the first image. you might be able to see the difference better if you enable points and make them large. if everything is translucent, you won't see a difference cause alpha-composition is commutative.

the last one has no difference for same reason, neither the grid nor the band has alpha 1.

dprokop replied:
Probably it's my color blindness then:)

ryantxu commented:
Rather than adding a new structure, can we use the existing thresholds field config to manage this?  We could have two additional boolean props on each threshold level: line and fill?  Similar to the field mapping popup we have discussed, perhaps the additional threshold settings are only available in a popup?

The other trick is handling the existing alerting threshold configs <-> field config threshold :thinking:

torkelo commented:
@ryantxu Rather than adding a new structure

We need the new structure anyway to control how the graph should render the thresholds (and if it should not), so some off / line / area / line + area option is needed. I do not think area fill & line opacity makes sense on the current universal threshold model as it would be very visualization/graph specific, hm...., tricky.  What we could do is make this graph threshold display option part of the thresholds option category. Right now it's in the Graph styles category. 

And my hope would be Grafana can control the line & area opacity (ie no option).  the old graph as "critical" and "warning" options where Grafana set the line and area colors. But there is a custom option where users can set whatever. Hm...

dprokop commented:
@torkelo @ryantxu 

I think threshold opacity could be controlled by the color of the threshold as the option allows rgba colors. I don't think we need an explicite option for setting the opacity. The thing is we would have to detect whether step's color is defined as a transparent color and skip the threshold color manipulation.

When it comes to new config option vs adding the control to the current thresholds model. My gut feeling is this brings huge possibilities, but the question is - are they really needed? I like the simplicity of @torkelo solution in this PR. Can't really see a case for mixing thresholds as lines and as fill. And if there is such a use case it must be a quite niche one.

torkelo commented:
> I think threshold opacity could be controlled by the color of the threshold as the option allows rgba colors. 

@dprokop then we won't be able to migrate (without changing the look) the old graph thresholds that have by default different line & area opacity in combination 

This is the default critical threshold in the old graph (here Grafana sets the line & area colors automatically, there is also a custom option where users can set both). 
![Screenshot from 2021-04-26 15-24-44](https://user-images.githubusercontent.com/10999/116089952-a69b2e80-a6a3-11eb-8244-523f40fbefc3.png)

My thinking was to keep this automatic behavior so users do not have to adjust the color opacity manually. And this way we can support line + area as well. But will be harder to migrate the custom thresholds that have a color that includes transparency (alpha), hm.....

leeoniya commented:
just a thought: it's pretty tough to discern between thresholds and actual bands, e.g. if the data is mostly flat. maybe in the future we can fill thresholds with a [CanvasPattern](https://developer.mozilla.org/en-US/docs/Web/API/CanvasPattern) instead of a series-alike fill.

dprokop commented:
> just a thought: it's pretty tough to discern between thresholds and actual bands or area fills, e.g. if the data is mostly flat. maybe in the future we can fill thresholds with a CanvasPattern instead of a series-alike solid fills.

@leeoniya for the a11y reasons we will need to support patterns (and custom shape points) in the Graph NG (https://github.com/grafana/grafana/issues/21120)

torkelo committed:
Merge remote-tracking branch 'origin/master' into graph-ng-thresholds

torkelo committed:
Merge remote-tracking branch 'origin/master' into graph-ng-thresholds

torkelo committed:
Merge remote-tracking branch 'origin/master' into graph-ng-thresholds

torkelo committed:
Updating theme for color picker and fixing issues

torkelo committed:
Updating theme for color picker and fixing issues

torkelo committed:
Minor progress

torkelo committed:
Added test dashboard

torkelo commented:
So using color with full opacity works for hiding a band. Think this makes more sense than a "Disable" or "hide" icon toggle button on the threshold, as a disabled/hidden threshold does not make much sense for many of the other panels. For this to work good I think we need a easy "transparent" color in the color picker. and special logic in threshold editor so you can see that there is color specified there, when a transparent color is selected you do not see the color picker anymore. I changed it so that the base color is used now, so this means that if you do not want the base (-infinity color) to show you have to change it to transparent. Find this better than very unintuitive logic that the base threshold was skipped. 

I also changed it so that if the color selected by the user has opacity then Grafana will not automatically set opacity, but if the color has no full alpha channel Grafana will automatically set a good opacity level. 

### Options
Moved the Show thresholds options into the thresholds category, it's at the top right now, should be at the bottom but we need a new option on field configs to control sorting/order for that. Do we need more options than Off | Line | Area ? maybe Line+Area? This option is still a custom graph option but lives in the standard thresholds category. 

### Color line
How to give users the option to color the line? We have two ways, either via color scheme or via this thresholds display mode. I am afraid of using the color scheme as it will be very complicated for the user (as it will require interactions between options in different categories). Using the new display mode and adding a new option "Color series", if set it could override the color scheme and use thresholds for the line color. But this is not great either, what to color the legend item with? This will not make the gradient color schemes usable for line colors either. 

Using color scheme: 
Color scheme: from thresholds

Maybe if this is set, we automatically use it for line colors, and if you want gradient you use the gradient mode option. so the color scheme from thresholds would be an orthogonal option to the graph threshold mode (off, line, area). 

 The trickiest thing now is how to control the option to color the line using thresholds, the most structural logic thing would be to enable users to set color scheme to "From thresholds"  but that 

![Screenshot from 2021-05-01 08-59-28](https://user-images.githubusercontent.com/10999/116774270-c774d280-aa5b-11eb-98d2-9452ff8c72e7.png)

torkelo committed:
Adding transparent and text colors to color picker and making them visible

torkelo commented:
Extending the color picker , adds transparent and text color options 
![Screenshot from 2021-05-01 17-28-07](https://user-images.githubusercontent.com/10999/116787194-1db93400-aaa3-11eb-8357-b97974a70655.png)

torkelo committed:
Good progress on migrations

ryantxu reviewed (commented):
This is looking good -- I think this gives us a good path forward.

There is something funky with handling max/min values -- I think the reported max is different than the canvas max so you are not rendering all the way to the next threshold:
![image](https://user-images.githubusercontent.com/705951/116802791-badb9300-aaca-11eb-800f-6940582b8859.png)
ie the blue/(purple!) threshold should not be shown at all, but 100 is the max.

ryantxu commented:
What do you think we should do about showing alerts on the TimeSeries panel?  I tried naively just enabling it, but it is more complex than that :)

In particular -- the angular version currently saves both an alert rule and threshold setting.  We should probably keep this and have the panel set thresholds on init (ie, not in a migration)

torkelo committed:
Fixed selected issue in color picker

torkelo commented:
struggling migrating thresholds definitions, I can migrate fill area transformations from old model that uses "less then" "greater then" rules,  for lines the mechanics would be different, the problem is that the new model always treats all thresholds as "greater then", and this works mostly fine, except when mixing line thresholds AND area. I can migrate a rule like "less then 20 red", as -Infinity red, 20 - transparent.  But I do not then get a red line at 20, as that is where the transparent region starts (becomes more trickier when multiple less then rules)

The problem is we have to handle these transparent thresholds in away that changes the meaning of the threshold. 

for example here:

![Screenshot from 2021-05-02 07-55-43](https://user-images.githubusercontent.com/10999/116804027-d7ec8200-ab1b-11eb-90a2-30901f816a5f.png)

This basically means we have to logically handle the transparent threshold a bit differently, that it really is an upper bound for the threshold below. 

![Screenshot from 2021-05-02 07-55-23](https://user-images.githubusercontent.com/10999/116804015-c73c0c00-ab1b-11eb-9b61-81b5aa27f385.png)

torkelo commented:
But that does not really work. The fundamental problem is these "less than" thresholds, the model only really supports "greater then" thresholds. I can express less than thresholds in a roundabout way but hard to know what the user intent is, and where to draw the lines.

torkelo commented:
Can we skip line thresholds and only support area thresholds? Line thresholds do not really work with our always "greater then" thresholds :(

torkelo committed:
Fixes

torkelo committed:
migration fix

torkelo committed:
Fixed test

torkelo committed:
Merge remote-tracking branch 'origin/master' into graph-ng-thresholds

torkelo commented:
Started working on alerting support for timeseries, and discovered lots of issues with annotations, and none of those changes impacted this so split that out into separate PR https://github.com/grafana/grafana/pull/33608

torkelo committed:
Fixing line rendering and refactoring

torkelo committed:
Minor ordering fix

torkelo committed:
fixed test

torkelo committed:
Merge branch 'master' of github.com:grafana/grafana into graph-ng-thresholds

torkelo committed:
Merge remote-tracking branch 'origin/master' into graph-ng-thresholds

torkelo committed:
Draw thresholds before axes & series and other tweaks

dprokop commented:
> Extending the color picker , adds transparent and text color options
> ![Screenshot from 2021-05-01 17-28-07](https://user-images.githubusercontent.com/10999/116787194-1db93400-aaa3-11eb-8357-b97974a70655.png)

@torkelo not sure what are these new options for?

dprokop commented:
> > Extending the color picker , adds transparent and text color options
> > ![Screenshot from 2021-05-01 17-28-07](https://user-images.githubusercontent.com/10999/116787194-1db93400-aaa3-11eb-8357-b97974a70655.png)
> 
> @torkelo not sure what are these new options for?

@torkelo ignore ^^ - i get it now when playing with thresholds settings

dprokop commented on packages/grafana-ui/src/components/uPlot/config.ts:
```diff
@@ -192,6 +192,24 @@ export interface StackingConfig {
+  Line = 'line',
+  Area = 'area',
+  LineAndArea = 'line+area',
+  Series = 'series',
```
What is the Series option?

torkelo replied:
when the line get's threshold colors

dprokop replied:
ah, I see. It's just not exposed in the UI, right?

torkelo replied:
not impemented yet

dprokop commented on packages/grafana-ui/src/components/uPlot/config/UPlotThresholds.ts:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    function addLines() {
+      // Thresholds below a transparent threshold is treated like "less then", and line drawn previous threshold
```
```suggestion
      // Thresholds below a transparent threshold is treated like "less than", and line drawn previous threshold
```

dprokop commented on public/app/plugins/panel/timeseries/migrations.ts:
```diff
@@ -312,6 +315,83 @@ export function flotToGraphOptions(angular: any): { fieldConfig: FieldConfigSour
+        });
+      }
+
+      if (threshold.op === 'lt') {
+        if (steps.length === 0) {
+          steps.push({
+            value: -Infinity,
+            color: getThresholdColor(threshold),
+          });
+        }
+
+        // next op is gt and there is a gap set color to transparent
+        if (next && next.op === 'gt' && next.value > threshold.value) {
+          steps.push({
+            value: threshold.value,
+            color: 'transparent',
+          });
+          // if next is a lt we need to use it's color
+        } else if (next && next.op === 'lt') {
+          steps.push({
+            value: threshold.value,
+            color: getThresholdColor(next),
+          });
+        } else {
+          steps.push({
+            value: threshold.value,
+            color: 'transparent',
+          });
+        }
+      }
+    }
```
ok, now i see what you've been fighting with :(

The transparent thresholds are quite unintuitive until you realize we treat thresholds as greater than :( Not sure how to explain that in the docs.  I think we gonna have plenty of wuestions about usage of these.

dprokop commented on packages/grafana-ui/src/components/ColorPicker/NamedColorsPalette.tsx:
```diff
@@ -25,15 +28,30 @@ export const NamedColorsPalette = ({ color, onChange, theme }: NamedColorsPalett
+          label="Transparent"
+          onClick={() => onChange('transparent')}
+        />
+        <ColorSwatch
+          isSelected={color === 'text'}
+          color={theme.colors.text.primary}
+          label="Text color"
+          onClick={() => onChange('text')}
+        />
```
Don't get why the text swatch is here. 

Another thing - since the color picker is used in few places - I am not sure having the transparen/text swatches everywhere makes sens. WDYT about having those only for the threshold's color picker - since they make sense there only ATM. 

Alternatively they can serve as "hide but preserve series" but I don't think this a way we want it to be used. 

And when you set the series color to transparent, the legend icon is not visible, making it hard to change the color back. Let's enable it for thresholds only?

torkelo replied:
> Don't get why the text swatch is here.

why not? it's a very useful color to pick, to get the text color independent of theme (think stat/gauge panel when you wan the value to have text color)

> Another thing - since the color picker is used in few places - I am not sure having the transparen/text swatches everywhere makes sens. WDYT about having those only for the threshold's color picker - since they make sense there only ATM. 

They are both very useful to stat panel as well

dprokop replied:
Ok, I get that. But in this case I'm looking from the POV of Legend color picker in which I think neither transparent nor text make much sense. That's why I'm proposing these additional color options to be optional and intentional rather than default for every color picker.

torkelo replied:
ah, maybe text color makes some sense but transparent not so much.

torkelo replied:
Yea agree, let's hide these options from legend color picker

dprokop reviewed (approved):
This is massive effort @torkelo, works really good.

torkelo committed:
Update packages/grafana-ui/src/components/uPlot/config/UPlotThresholds.ts

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>
